src/v/raft/replicate_batcher.cc

```diff
@@ -63,7 +67,7 @@ replicate_stages replicate_batcher::replicate(
 ss::future<> replicate_batcher::stop() {
     // we keep a lock here to make sure that all inflight requests have finished
     // already
-    return _lock.with([this]() {
+    return _bg.close().then([this] {
```

Comment:
it doesn't seem like switching out the gate for a batcher-only gate is enough to be safe. for example, batcher do_flush is called directly by consensus (via replicate configuration) an do flush runs stm->apply before taking the batcher gate. so if the batcher is closed while stm->apply is running then the gate will be closed while stm->apply is running. I'm assuming there is a shutdown relationship between consensus and the batcher such that after the batcher gate closes consensus assumes batcher holds no more references, etc... so that basic, common contract seems like it is violated. I didn't look at all the other entry points.

Reply:
This will not break the dependency chain since `do_flush` doesn't use any of the `batcher` resources. I agree that the interfaces here are not perfect. Will try to refactor this to make it more elegant and clear.

Reply:
> This will not break the dependency chain since `do_flush` doesn't use any of the `batcher` resources.

it captures `this` without a gate...

but my concern wasn't the batcher resources, it was with consensus resources and stopping the batcher implied wrt to the consensus shutdown process. it looks like the consensus gate protects access to batcher::do_flush, but that isn't obvious at the interface stopping interface level.

Reply:
I refactored this, now consensus doesn't use batcher `do_flush` method